pmda/density.py

```diff
@@ -0,0 +1,158 @@
+        self._atomselection = atomselection
+        self._metadata = metadata
+        self._padding = padding
+        self._updating = updating
+        self._parameters = parameters
+        self._gridcenter = gridcenter
+        self._xdim = xdim
+        self._ydim = ydim
+        self._zdim = zdim
+        self._trajectory = u.trajectory
+        self._n_frames = u.trajectory.n_frames
+        if updating and atomselection is None:
+            raise ValueError("updating=True requires a atomselection string")
+        elif not updating and atomselection is not None:
+            raise ValueError("""With updating=False, the atomselection='{}' is
+                        not used and should be None""".format(atomselection))
+
+    def _prepare(self):
+        coord = self.current_coordinates(self._atomgroup, self._atomselection,
+                                         self._updating)
+        box, angles = (self._trajectory.ts.dimensions[:3],
+                       self._trajectory.ts.dimensions[3:])
+        if self._gridcenter is not None:
+            # Generate a copy of smin/smax from coords to later check if the
+            # defined box might be too small for the selection
+            smin = np.min(coord, axis=0)
+            smax = np.max(coord, axis=0)
+            # Overwrite smin/smax with user defined values
+            smin, smax = _set_user_grid(self._gridcenter, self._xdim,
+                                        self._ydim, self._zdim, smin, smax)
+        else:
+            # Make the box bigger to avoid as much as possible 'outlier'. This
+            # is important if the sites are defined at a high density: in this
+            # case the bulk regions don't have to be close to 1 * n0 but can
+            # be less. It's much more difficult to deal with outliers.  The
+            # ideal solution would use images: implement 'looking across the
+            # periodic boundaries' but that gets complicated when the box
+            # rotates due to RMS fitting.
+            smin = np.min(coord, axis=0) - self._padding
+            smax = np.max(coord, axis=0) + self._padding
+            BINS = fixedwidth_bins(self._delta, smin, smax)
+            arange = np.transpose(np.vstack((BINS['min'], BINS['max'])))
+            bins = BINS['Nbins']
+            # create empty grid with the right dimensions (and get the edges)
+            grid, edges = np.histogramdd(np.zeros((1, 3)), bins=bins,
+                                         range=arange, normed=False)
+            grid *= 0.0
+        self._grid = grid
+        self._edges = edges
+        self._arange = arange
+        self._bins = bins
+
+    def _single_frame(self, ts, atomgroups):
+        coord = self.current_coordinates(atomgroups[0], self._atomselection,
+                                         self._updating)
+        h, edges = np.histogramdd(coord, bins=self._bins, range=self._arange,
+                                  normed=False)
```

Comment:
`edges` is not used in other place. Try to look at the out put of `np.histogramdd` and find a way to get rid of `edges`.